pandas/tests/indexes/test_base.py

```diff
@@ -201,6 +201,25 @@ def __array__(self, dtype=None):
             result = pd.Index(ArrayLike(array))
             tm.assert_index_equal(result, expected)
 
+    def test_constructor_int_dtype_float(self):
+        # GH 18400
+        data = [0., 1., 2., 3.]
+
+        expected = Int64Index([0, 1, 2, 3])
+        result = Index(data, dtype='int64')
+        tm.assert_index_equal(result, expected)
+
+        expected = UInt64Index([0, 1, 2, 3])
+        result = Index(data, dtype='uint64')
+        tm.assert_index_equal(result, expected)
+
+        # fall back to Float64Index
+        data = [0.0, 1.1, 2.2, 3.3]
+        expected = Float64Index(data)
```

Comment:
In light of Case III in #15832, do we actually want this behavior, or should it really coerce to `Int64Index([0, 1, 2, 3])`?  Originally wrote this test case to make sure all relevant paths were hit, without seeing #15832.

Reply:
these should raise as they are passing a dtype that is non convertible

Reply:
Removed the `Float64Index` portion of the test.  Was able to get it to raise, but ended up breaking other things.  Not immediately sure of the fix, but seems outside the scope of this PR, though can look into it more if need be. None of the code I wrote actually depends on that behavior; originally included it to make sure all code paths were tested.